Dynamic scale factor updates

last-partizan opened the pull request:
This is draft attemmt to support `ScaleFactorChanged` event.

I use target branch `fix-scaling`, to show only relevant changes in this branch. I'll change it to main when `fix-scaling` is merged.

Here i refactored CachingShaper to always include `options`, and not `Option<FontOptions>`. And removed Option from few other parts, becouse i was lost in `unwrap`s and `map`s. And this is probably works as expected.

Font size is changing when i change scaling in display settings, but window size does not.

Please take a look at this, try it, maybe you have some ideas for how to make window scale too.

Please, pay close attention to any rust-specific features i try to use here, (like pointers, and dereferencing), i usually work with python, and here i write code with help from langserver, rust-compiler and google.

## What kind of change does this PR introduce?
- Fix
- Refactor

## Did this PR introduce a breaking change?
- No

smolck commented on src/renderer/fonts/font_loader.rs:
```diff
@@ -54,6 +54,16 @@ pub struct FontKey {
 }
 
+impl FontKey {
+    pub fn default() -> FontKey {
```
There’s a trait for this, see https://doc.rust-lang.org/std/default/trait.Default.html:

```rust
impl Default for FontKey {
    fn default() -> Self {
      // …
    }
}
```

last-partizan replied:
Thanks!

smolck commented:
> Here i refactored CachingShaper to always include options, and not Option<FontOptions>.

Why did you do this? I’m not familiar with the codebase enough to know why it was an Option type in the first place, so not sure if it’s really unnecessary or not . . .

> And removed Option from few other parts, becouse i was lost in unwraps and maps. And this is probably works as expected.

If they really are unnecessary, that’s fine. But in some cases having a `None` value is a valid thing, in which case you’ll want to leave it as Option. But I haven’t really studied the code at this point so this may not be relevant here.

last-partizan commented:
>> Here i refactored CachingShaper to always include options, and not Option.
>
> Why did you do this? I’m not familiar with the codebase enough to know why it was an Option type in the first place, so not sure if it’s really unnecessary or not . . .

>>  And removed Option from few other parts, becouse i was lost in unwraps and maps. And this is probably works as expected.
> 
> If they really are unnecessary, that’s fine. But in some cases having a None value is a valid thing, in which case you’ll want to leave it as Option. But I haven’t really studied the code at this point so this may not be relevant here.


Having default values instead of Option makes a lot of code unnecessary, here is good example:

```diff
-        self.options
-            .as_ref()
-            .map(|options| options.size)
-            .unwrap_or(DEFAULT_FONT_SIZE)
-            * self.scale_factor
+        self.options.size * self.scale_factor
```

And if unwrapped value has optional field, that's one more level of unwraps or maps.

Kethku commented on src/renderer/fonts/caching_shaper.rs:
```diff
@@ -68,26 +56,30 @@ impl CachingShaper {
 
     pub fn update_font(&mut self, guifont_setting: &str) -> bool {
-        let new_options = FontOptions::parse(guifont_setting, DEFAULT_FONT_SIZE);
+        println!("Updating font: {}", guifont_setting);
```
use a trace here rather than println

last-partizan replied:
That's temporary code until this is draft. I'll remove all prints after this is working as expected.

Kethku commented on src/renderer/fonts/caching_shaper.rs:
```diff
@@ -68,26 +56,30 @@ impl CachingShaper {
+
+    fn reset_font_loader(&mut self) {
+        let font_size = self.options.size * self.scale_factor;
+        println!("Using font_size: {:.2}", font_size);
```
same here. Trace rather than log

Kethku commented on src/renderer/mod.rs:
```diff
@@ -87,12 +87,24 @@ impl Renderer {
+        let (font_width, font_height) = self.shaper.font_base_dimensions();
+        self.font_width = font_width;
+        self.font_height = font_height;
+        println!(
```
also here. Trace rather than println (note: i realize this is probably for testing, but is likely useful for debugging purposes, so marking these in the review)

Kethku replied:
```suggestion
        trace!(
```

Kethku commented on src/window/window_wrapper/mod.rs:
```diff
@@ -159,6 +165,7 @@ impl GlutinWindowWrapper {
         let previous_size = self.saved_inner_size;
 
         if previous_size != current_size {
+            println!("Updating grid size: {:#?}", current_size);
```
and here: trace rather than println

Kethku replied:
```suggestion
            trace!("Updating grid size: {:#?}", current_size);
```

last-partizan commented on .github/workflows/build.yml:
```diff
@@ -163,11 +167,15 @@ jobs:
-        name: neovide-linux
-        path: ./target/release/neovide
+        name: neovide-linux.tar.gz
+        path: ./target/release/neovide.tar.gz
```
Hmm, i don't understand why are this changes here, i just synced branch with master :(

Please ignore this, i'll resolve it later.

shaunsingh replied:
Linux/windows builds are now distriubuted in a .tar.gz format as changed in https://github.com/Kethku/neovide/pull/822

last-partizan replied:
That was becouse target branch was fix-scaling instead of main. Resolved.

clason commented:
Just to mention this here explicitly: the `fix-scaling` PR "broke" HiDPI (at least macOS Retina) screens; it looks like the internal pixel doubling is applied twice -- the font is now huge! (But renders fine at that large size, hence the scare quotes around "broke".)

I assume correct handling of Retina screens is part of this PR?

last-partizan commented:
@clason this PR does not touch scaling itself, only dynamic changing of scale factor.

after `fix-scaling` font size was expected to get bigger, becouse we converted points to pixels. What is your font size for neovide?

on linux before `fix-scaling` my font-size was 14 (but actually that was 11pt).

maybe on mac-os points and pixels are the same. If your font size was consistent with other apps, i'll make another PR to fix that.

clason commented:
Ah, I see, then this was my misconception. I don't configure Neovide at all, so it must be using some defaults (which ones?) that are less than reasonable on a HiDPI screen ;)

last-partizan commented:
You could run `neovide --log`, and then `grep 'scale' neovide_rCURRENT.log`

It should look like this:
```
INFO [neovide::window::window_wrapper] window created (scale_factor: 1.6041666666666667, font_size: 18x37)
```

(that's my 11pt, multiplied by 1.3 and with some extra height added)

clason commented:
```
INFO [neovide::window::window_wrapper] window created (scale_factor: 2, font_size: 23x46)
```
(macOS, external monitor at 2560x1440)

last-partizan commented:
I replaced all "println" with "trace", and fixed bug with incorrect scaling.

@Kethku please check how it works for you.

Kethku commented:
This works perfectly on my machine. Merging.

Kethku commented:
Thanks for pushing this forward! Your changes have made a big difference for the stability of the app
